File: python/Iterator/waitress.py
from menu import PancakeHouseMenu


class Waitress:
    def __init__(self, pancake_house_menu: PancakeHouseMenu, diner_menu) -> None:
        self.pancake_house_menu = pancake_house_menu
        self.diner_menu = diner_menu

    def printMenu(self) -> None:
        print("Menu")
        print("-----")
        print("BREAKFAST")
        self.__printMenu(self.pancake_house_menu)
        print("")
        print("LUNCH")
        self.__printMenu(self.diner_menu)

    def __printMenu(self, menu) -> None:
        for m in menu.createIterator():
            print(f"{m.name}, {m.price} -- {m.description}")

File: c/Iterator/waitress.c
#include "waitress.h"

static void _printMenu(Menu *menu);

void WaitressNew(Waitress *self, PancakeHouseMenu *p_menu, DinerMenu *d_menu)
{
    self->p_menu = p_menu;
    self->d_menu = d_menu;
    self->printMenu = Waitress_printMenu;
}
void Waitress_printMenu(Waitress *self)
{
    printf("Menu\n");
    printf("-----\n");
    printf("BREAKFAST\n");
    _printMenu((Menu *)self->p_menu);
    printf("\n");
    printf("LUNCH\n");
    _printMenu((Menu *)self->d_menu);
}

static void _printMenu(Menu *menu)
{
    Iterator_MenuItem iter = menu->createIterator(menu);
    while (iter.hasNext(&iter))
    {
        MenuItem item = iter.next(&iter);
        printf("%s, %.2f -- %s\n", item.name, item.price, item.description);
    }
}